src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.Blocking.cs

```diff
@@ -0,0 +1,384 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Threading
+{
+    internal sealed partial class PortableThreadPool
+    {
+        public short MinThreadsGoal
+        {
+            get
+            {
+                _threadAdjustmentLock.VerifyIsLocked();
+                return Math.Min(_separated.numThreadsGoal, TargetThreadsGoalForBlockingAdjustment);
+            }
+        }
+
+        private short TargetThreadsGoalForBlockingAdjustment
+        {
+            get
+            {
+                _threadAdjustmentLock.VerifyIsLocked();
+
+                short targetThreadsGoal = _minThreads;
+                if (_numBlockedThreads <= 0)
+                {
+                    return targetThreadsGoal;
+                }
+
+                short maxThreads = MaxThreadsForBlockingAdjustment;
+                targetThreadsGoal += _numBlockedThreads;
+                if (targetThreadsGoal < _numBlockedThreads || targetThreadsGoal > maxThreads)
+                {
+                    targetThreadsGoal = maxThreads;
+                }
+                return targetThreadsGoal;
+            }
+        }
+
+        private short MaxThreadsForBlockingAdjustment
+        {
+            get
+            {
+                _threadAdjustmentLock.VerifyIsLocked();
+
+                short result = (short)(_minThreads + BlockingConfig.MaxThreadsToAddBeforeFallback);
+                return result < BlockingConfig.MaxThreadsToAddBeforeFallback || result > _maxThreads ? _maxThreads : result;
+            }
+        }
+
+        public void NotifyThreadBlocked()
+        {
+            if (!BlockingConfig.IsCooperativeBlockingEnabled || !Thread.CurrentThread.IsThreadPoolThread)
+            {
+                return;
+            }
+
+            bool wakeGateThread = false;
+            _threadAdjustmentLock.Acquire();
+            try
+            {
+                _numBlockedThreads++;
+                Debug.Assert(_numBlockedThreads > 0);
+
+                if (_pendingBlockingAdjustment != PendingBlockingAdjustment.WithDelayIfNecessary &&
+                    _separated.numThreadsGoal < TargetThreadsGoalForBlockingAdjustment)
+                {
+                    if (_pendingBlockingAdjustment == PendingBlockingAdjustment.None)
+                    {
+                        wakeGateThread = true;
+                    }
+                    _pendingBlockingAdjustment = PendingBlockingAdjustment.WithDelayIfNecessary;
+                }
+            }
+            finally
+            {
+                _threadAdjustmentLock.Release();
+            }
+
+            if (wakeGateThread)
+            {
+                GateThread.Wake(this);
+            }
+        }
+
+        public void NotifyThreadUnblocked()
+        {
+            if (!BlockingConfig.IsCooperativeBlockingEnabled || !Thread.CurrentThread.IsThreadPoolThread)
```

Comment:
Should we have NotifyThreadBlocked return a bool and then only call NotifyThreadUnblocked if it returned true?  Then this if block could be removed, and we'd avoid the extra TLS access for CurrentThread.  Not a big deal, though; this is already a more expensive code path.

Reply:
Good idea, will fix